services/navigation/builtin/builtin.go

```diff
@@ -283,26 +280,42 @@ func (svc *builtIn) startWaypoint(extra map[string]interface{}) error {
 				// - Remember that we added -1*bearingDelta instead of steeringDir
 				// - Test both naval/land to prove it works
 				if err := svc.base.Spin(ctx, -1*bearingDelta, svc.degPerSec, nil); err != nil {
-					return fmt.Errorf("error turning: %w", err)
+					return errors.Wrap(err, "failed while turning")
 				}
 
 				distanceMm := distanceToGoal * 1000 * 1000
 				distanceMm = math.Min(distanceMm, 10*1000)
 
 				// TODO: handle swap from mm to meters
 				if err := svc.base.MoveStraight(ctx, int(distanceMm), (svc.metersPerSec * 1000), nil); err != nil {
-					return fmt.Errorf("error moving %w", err)
+					return errors.Wrap(err, "failed while moving straight")
 				}
 
 				return nil
 			}
 
-			if err := navOnce(svc.cancelCtx); err != nil {
-				svc.logger.Infof("error navigating: %s", err)
+			if cancelCtx.Err() != nil {
+				return
+			}
+
+			svc.mu.Lock()
+			navOnceCancelCtx, fn := context.WithCancel(cancelCtx)
+			svc.navOnceCancelFunc = fn
+			svc.mu.Unlock()
+
+			if err := navOnce(navOnceCancelCtx); err != nil {
+				if svc.waypointIsDeleted() {
+					svc.logger.Info("skipping waypoint since it was deleted")
+					continue
+				}
+
+				svc.logger.Infof("skipping waypoint due to error while navigating towards it: %s", err)
+				if err := svc.waypointReached(cancelCtx); err != nil {
+					svc.logger.Info("can't mark waypoint as reached, exiting navigation due to error: %s", err)
```

Comment:
This isn't actually existing navigation, is it? It's just going to go to the next waypoint in the loop.

Reply:
Good catch! This should return.